questionpy_server/cache.py

```diff
@@ -111,34 +118,35 @@ def put(self, key: str, value: bytes) -> Path:
             raise SizeError(f"Item itself exceeds maximum allowed size of {self.max_bytes} bytes",
                             max_size=self.max_bytes, actual_size=size)
 
-        # Update `_total_bytes` depending on whether the key existed already or not.
-        if key in self._files:
-            self._total_bytes -= self._files[key].size
-        self._total_bytes += size
+        async with self._lock:
+            # Update `_total_bytes` depending on whether the key existed already or not.
+            if key in self._files:
+                self._total_bytes -= self._files[key].size
+            self._total_bytes += size
 
-        # Save the bytes on filesystem.
-        path = self.directory / (key + self._extension)
+            # Save the bytes on filesystem.
+            path = self.directory / (key + self._extension)
 
-        if not self.directory.is_dir():
-            self.directory.mkdir(parents=True, exist_ok=True)
+            if not self.directory.is_dir():
+                self.directory.mkdir(parents=True, exist_ok=True)
 
-        if size != path.write_bytes(value):
-            raise IOError("Failed to write bytes to file")
+            if size != path.write_bytes(value):
```

Comment:
Das muss noch async gemacht werden. Achtung: Wenn der Prozess abschmiert, während ein Teil der Datei schon geschrieben wurde, dann weiß beim nächsten Start diese Klasse nicht, dass die Datei unvollständig ist.

Du solltest daher den Inhalt zunächst in eine temporäre Datei schreiben und die Datei am Ende umbenennen. Wenn `__init__` temporäre Dateien findet, soll es diese löschen.

Und noch was:  `self._total_bytes` sollte besser erst verändert werden, nachdem die Datei erfolgreich am endgültigen Pfad gespeichert wurde.